### ISSUE TITLE
Write documentation for embedded systems

### ISSUE THREAD

**Issue by shrit** [mlpack/mlpack#2922]
Hello everyone,

I open this issue in order to discuss the documentation form for the embedded system integration in mlpack. Once #2531 is merged, users will be initially able to cross-compile mlpack for aarch64 embedded systems, however, to do this we need to discuss the documentation form in which we will present this feature.

There are several forms in doing this such as:

1) Writing a tutorial for users on mlpack, where? Website, Github, Wiki, Readme, 
2) Adding documentation instead of tutorials, where?
3) Other thoughts?

I do think we need to mention that mlpack is now compatible with embedded systems, such as writing keywords or phrases on Readme, mlpack front page website.

I am open to any suggestion, let me know what do you think, I would vote for a tutorial that the user can read on mlpack website.

**Comment by rcurtin** [comment]
I definitely agree that we should rework our documentation and README to point to our support for embedded devices and cross-compilation.

I think maybe the easiest way to go, for now, is to put a tutorial in `doc/guide/` kind of like `build.hpp` but specifically for cross-compilation.

It would be great too if we could modify the `examples/` repository to provide some information on how a user might compile those examples (or at least the C++ ones) to an embedded device.

Anyway, those are just some ideas.

**Comment by shrit** [comment]
Agreed, considering the `example` repository, maybe we should provide a template for a cross-compilation makefile?

**Comment by rcurtin** [comment]
Yeah, that could be a nice addition!

**Comment by zoq** [comment]
I would start with a guide, and keep it simple something like https://www.tensorflow.org/lite/guide/build_cmake_arm. And then maybe add a tutorial that shows how to run a specific example on a rpi or something similar. I'm not super happy with how the current tutorials are embedded, looks like a medium post nowadays get's a lot more exposure.

**Comment by shrit** [comment]
@zoq Agreed, maybe we need to advertise better the tutorial on mlpack website.
It might be a good idea to have a medium post, but it should be additional and should not replace the mlpack tutorial on mlpack website.
I find mlpack tutorials are good, well explained, and easy to find, much easier than TensorFlow. We only need alot of them

**Comment by zoq** [comment]
My suggestion would be more like having each tutorial on medium as well as on the mlpack homepage. I'm certain you get more page impressions on the medium post as on the tutorials, if medium can do one thing it's advertising, and that's not something we can get if we post this on hn or have a more prominent position on the mlpack start page.

Also, I agree the tutorials are good, I said the embedding is what I think isn't optimal, like you can't link sections. What I like about the TensorFlow tutorials is that it's more like a notebook, you have the code snippets you can copy with a single click and you have the output directly beneath.

There is an open issue that suggested to switch from the doxygen generate tutorials to Sphinx, which would also allow us to embed code snippets, anyway my suggestion would be to write the tutorials in a way that they can be read by medium and doxygen, that probably means markdown.

**Comment by shrit** [comment]
@zoq Totally agreed, I prefer sphinx over doxygen, but I admire the most is gitbook
In `gitbook`, you can write everything in markdown, and it is even better than sphinx or doxygen.
Gitbook is easier to read, cleaner, and has the same logic to copy-paste code snippets.
Let me know what do you think :+1:

**Comment by zoq** [comment]
I like the `gitbook` style as well, but if I remember right you can't host that yourself?

**Comment by shrit** [comment]
Yes, you can host it on your own website, there is no issue with this. You have only to install it https://github.com/GitbookIO/gitbook/blob/master/docs/setup.md

**Comment by rcurtin** [comment]
I'd love to transition our tutorials to Sphinx (and our documentation too!) but it would be a good bit of work.  #1877 discussed it; maybe it is actually getting close to time to do it. :)

I think @zoq is right that Medium might be a far more effective place to post these tutorials or examples.  (We could also keep them on our website, too.)

**Comment by heisenbuug** [comment]
I agree about the medium article, actually, I was planning to write a medium article about `Getting started with Open Source Development with Mlpack` where I was planning to write about how to choose your first issue and how to get started with solving bugs, let me know if you guys have any suggestions. 

Also, I explored sphinx and doxygen a bit, I will get familiar with gitbook too. I made a small sample project with `doxygen+sphinx+breathe` just to mess around, I am new to cmake and build systems so its taking some time to figure it out...

**Comment by Aakash-kaushik** [comment]
@zoq, @rcurtin and @shrit tagging you all so we can continue our discussion for documentation here and make a plan to proceed.

**Comment by shrit** [comment]
@Aakash-kaushik Thanks, I would prefer to open a separate issue that is entitled `improve documentation or mlpack` or something similar, and tag this one there as a first initiative because when documentations are written for embedded systems this will be closed.
Then each one can comment with improvement and you can edit the main list of available things to write and focus on the same way we have done for `ann-vtable`
Let me know what do you think

**Comment by Aakash-kaushik** [comment]
@shrit seems like a good plan but i am not much aware about the documentation to write a proper guide on what to avoid and what not to and might not be the best person to help write up the issue i will help with it though.
